### PR TITLE
Sort symbols by address, then parentage, then index

### LINKED ISSUE #1151: Sort local symbols before global ones in sym files
Sym files "may or may not be sorted by location" [according to the spec](https://rgbds.gbdev.io/sym/), and the [rgblink docs](https://rgbds.gbdev.io/docs/v0.6.1/rgblink.1/) do not mention sorting. The current implementation claims "Symbols are ordered by address, or else by original index for a stable sort". It may be preferable to sort local labels differently from global ones, so debuggers will be more likely to show the correct symbol when they share an address.

Action:

```
$ cat foo.asm
SECTION "test2", ROM0[3]
Quux:
 ret
 .end
SECTION "test", ROM0[0]
Foo:
 ret
 .end
Bar:
 ret
 .end
Baz:
 ret
 .end
 
$ rgbasm -o foo.o foo.asm
$ rgblink -n foo.sym foo.o
$ cat foo.sym
```

Current output:

```
; File generated by rgblink
00:0000 Foo
00:0001 Foo.end
00:0001 Bar
00:0002 Baz
00:0002 Bar.end
00:0003 Baz.end
00:0003 Quux
00:0004 Quux.end
```

Desired output:

```
; File generated by rgblink
00:0000 Foo
00:0001 Foo.end
00:0001 Bar
00:0002 Bar.end
00:0002 Baz
00:0003 Baz.end
00:0003 Quux
00:0004 Quux.end
```

### PR DESCRIPTION
**Pull request by Rangi42**
Fixes #1151

With this file:

```asm
SECTION "test3", ROM0
Eggs:
 ret
 .end
Egg:
 ret
 .end
SECTION "test2", ROM0[3]
Quux:
 ret
 .end
Frob:
 .frob:
SECTION "test", ROM0[0]
Foo:
 ret
 .end
Bar:
 ret
 .end
Baz:
 ret
 .end
```

The previous .sym output was this:

```
00:0000 Foo
00:0001 Foo.end
00:0001 Bar
00:0002 Baz
00:0002 Bar.end
00:0003 Baz.end
00:0003 Quux
00:0004 Frob
00:0004 Frob.frob
00:0004 Quux.end
00:0004 Eggs
00:0005 Egg
00:0005 Eggs.end
00:0006 Egg.end
```

And the fixed .sym output is this:

```
00:0000 Foo
00:0001 Foo.end
00:0001 Bar
00:0002 Bar.end
00:0002 Baz
00:0003 Baz.end
00:0003 Quux
00:0004 Frob
00:0004 Frob.frob
00:0004 Quux.end
00:0004 Eggs
00:0005 Eggs.end
00:0005 Egg
00:0006 Egg.end
```

(Any particular sorting is not guaranteed by the man page, and .sym output is not something we've previously tested, so this isn't one of the CI test cases; just something to demonstrate the change for review.)

### CODE REVIEW

File: src/link/output.c [hunk 1]
```diff
@@ -334,6 +335,24 @@ static int compareSymbols(void const *a, void const *b)
...
+	if (sym1_local != sym2_local) {
+		size_t sym1_len = strlen(sym1_name);
+		size_t sym2_len = strlen(sym2_name);
+
+		// Sort parent labels before their child local labels
+		if (!strncmp(sym1_name, sym2_name, sym1_len) && sym2_name[sym1_len] == '.')
+			return -1;
+		if (!strncmp(sym2_name, sym1_name, sym2_len) && sym1_name[sym2_len] == '.')
+			return 1;
+		// Sort local labels before unrelated global labels
```
Comment (ISSOtm) [inline src/link/output.c:352]: I think I see the sentiment, but unfortunately this doesn't play nice with #1157, where "early" local labels will be sorted wrong. (Likewise the initial locality inequality check.)

Should we bite the bullet and claim those as rare anyway?

Comment (Rangi42) [reply src/link/output.c:352]: You mean if `Foo` has the same address as `Bar.local`, both of which come before `Bar` itself, then `Bar.local` will be sorted first? That is a rare case, and tbh I think it's remarkable enough to be worth prioritizing the weird local-before-main label in an emulator disassembly view.

Comment (ISSOtm) [reply src/link/output.c:352]: What about

```asm
Foo:
    db 42
.end ; Foo.end
Bar:
    db 69
.end ; Bar.end
```

Comment (Rangi42) [reply src/link/output.c:352]: Both current rgblink, and rgblink with this PR, output the same thing:

```
$ cat <<EOF | rgbasm -o - - | rgblink -n - -
section "test", rom0[0]
Foo: db 42
.end
Bar: db 69
.end
EOF
; File generated by rgblink
00:0000 Foo
00:0001 Foo.end
00:0001 Bar
00:0002 Bar.end
```

**Review by ISSOtm — APPROVED** [review]

File: src/link/output.c (hunk 1 above)
Comment (ISSOtm) [reply src/link/output.c:352]: The breakage isn't new, and minor. Let's go with it.

Comment (Rangi42) [reply src/link/output.c:352]: If you consider the early local label sorting behavior "breakage" (it's a new feature which already worked that way), I'll file a separate issue.

**Pull request merged**